// five-server/lib/public.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PREVIEW_FULLSCREEN = exports.PREVIEW = exports.PHP_TEMPLATE = exports.PHP_ERROR = exports.STATUS_CODE = exports.INJECTED_CODE = void 0;
const path_1 = require("path");
const fs_1 = require("fs");
exports.INJECTED_CODE = (0, fs_1.readFileSync)((0, path_1.join)(__dirname, '../client/injected.js'), 'utf8');
exports.STATUS_CODE = (0, fs_1.readFileSync)((0, path_1.join)(__dirname, '../public/serve-preview/status.html'), 'utf8');
exports.PHP_ERROR = (0, fs_1.readFileSync)((0, path_1.join)(__dirname, '../public/serve-php/error.html'), 'utf8');
exports.PHP_TEMPLATE = (0, fs_1.readFileSync)((0, path_1.join)(__dirname, '../public/serve-php/index.html'), 'utf8');
exports.PREVIEW = (0, fs_1.readFileSync)((0, path_1.join)(__dirname, '../public/serve-preview/preview.html'), 'utf8');
exports.PREVIEW_FULLSCREEN = (0, fs_1.readFileSync)((0, path_1.join)(__dirname, '../public/serve-preview/preview-fullscreen.html'), 'utf8');
